shell/platform/android/io/flutter/plugin/editing/ListenableEditingState.java

```diff
@@ -179,6 +257,54 @@ public SpannableStringBuilder replace(
       Log.e(TAG, "editing state should not be changed in a listener callback");
     }
 
+    Log.e(
+        "DELTAS",
+        "replace(" + start + ", " + end + ", " + tb + ", " + tbstart + ", " + tbend + ")");
+
+    final boolean isDeletionGreaterThanOne = end - (start + tbend) > 1;
+    final boolean isCalledFromDelete = tb == "" && tbstart == 0 && tbstart == tbend;
+
+    final boolean isReplacedByShorter = isDeletionGreaterThanOne && (tbend - tbstart < end - start);
+    final boolean isReplacedByLonger = tbend - tbstart > end - start;
+    final boolean isReplacedBySame = tbend - tbstart == end - start;
+
+    // Is deleting/inserting at the end of a composing region.
+    final boolean isDeletingInsideComposingRegion = !isReplacedByShorter && start + tbend < end;
+    final boolean isInsertingInsideComposingRegion = start + tbend > end;
+
+    // To consider the cases when autocorrect increases the length of the text being composed by
+    // one, but changes more than one character.
+    final boolean isOriginalComposingRegionTextChanged =
+        (isCalledFromDelete || isDeletingInsideComposingRegion || isReplacedByShorter)
+            || !toString()
+                .subSequence(start, end)
+                .equals(tb.toString().subSequence(tbstart, end - start));
+
+    final boolean isEqual =
+        toString().subSequence(start, end).equals(tb.toString().subSequence(tbstart, tbend));
+
+    // A replacement means the original composing region has changed, anything else will be
+    // considered an insertion.
+    final boolean isReplaced =
+        isOriginalComposingRegionTextChanged
+            && (isReplacedByLonger || isReplacedBySame || isReplacedByShorter);
+
+    if (isEqual) {
+      Log.e("DELTAS", "EQUALITY");
+      setDeltas(toString(), "", "EQUALITY", -1, -1);
```

Comment:
Would it be better to just not call updateEditingStateWithDelta at all in this case?

Reply:
Yes, but it is called every time the editing state changes see https://github.com/flutter/engine/pull/28175/files#diff-6d16487c29b469c59861d81341897ed392331e365647f6c13c18459800cc854dR653 I think this is the place we want to put it to keep it in sync with the actual text editing changes.

The editing state could change for reasons not pertaining to the actual raw text chagning, such as a change in the selection/composing region, so I think we need this `equality` case for these cases.

Reply:
Ah so if the user just moves the selection around with the arrow keys, they will get a bunch of EQUALITY deltas.

I wonder how exactly these calls would be used in the framework, similar to my other comment [here](https://github.com/flutter/engine/pull/28175#discussion_r692405158), to help me see how these equality deltas would be used.

Reply:
A use case of this could be.

Developer subscribes to `updateEditingValueWithDelta`. They update the `currentTextEditingValue` by applying the delta to it. 

Say we are typing "hello"
Engine sends "h" delta
Framework receives "h" delta applies it to `currentTextEditingValue`.
Engine sends "e" delta
Framework receives "e" delta and applies it to `currentTextEditingValue`.
.
.
.
.
Engine sends "o" delta
Framework receives "o" delta and applies it to `currentTextEditingValue`.
User changes selection
Engine sends equality delta to framework
Framework receives equality delta and does nothing (or updates their current selection/composing region with the new one).

Writing this makes me think that we should also be sending the new selection/composing region with this delta. So when a developer gets a delta from `updateEditingValueWithDelta` they also get the updated selection/composing region they can apply to the `TextEditingValue`/whatever structure they have representing a `TextEditingValue`. Any thoughts on this?